clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -1,34 +1,46 @@
-// RUN: %clang_cc1 %s -fsyntax-only -ast-dump -fsycl-is-device -triple spir64 -verify
+// RUN: %clang_cc1 %s --std=c++17 -fsyntax-only -fsycl-is-device -Wno-unused -verify
+// RUN: %clang_cc1 %s --std=c++17 -fsyntax-only -DTRIGGER_ERROR -fsycl-is-device -Wno-unused -verify
+
+#ifdef TRIGGER_ERROR
 
-// expected-error 8 @+2{{kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
 template <typename KernelName, typename KernelType>
-[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) {
+[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) { // expected-error 6 {{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
   kernelFunc();
 }
-template <template <typename> typename Outer, typename Inner>
-struct S {
-  void operator()() const;
-};
-
-template <typename Ty>
-struct T {};
 
 template <typename Func>
 void kernel3func(const Func &F3) {
-  // expected-note@+1{{__builtin_unique_stable_name evaluated here}}
+  // note@+1{{'__builtin_unique_stable_name' evaluated here}}
   constexpr const char *F3_output = __builtin_unique_stable_name(F3);
-  // expected-error@+1{{error: kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+  // error@+1{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{in instantiation of function template specialization}}
   kernel_single_task<class kernel3>(F3);
 }
 
 template <typename Func>
 void kernel4func(const Func &F4) {
-  // expected-note@+1{{__builtin_unique_stable_name evaluated here}}
+  // note@+1{{'__builtin_unique_stable_name' evaluated here}}
   constexpr const char *F4_output = __builtin_unique_stable_name(F4);
-  // expected-error@+1{{error: kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+  // error@+1{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{in instantiation of function template specialization}}
   kernel_single_task<class kernel4>([]() {});
 }
 
+#else
+// This section is not supposed to emit any diagnostics but there are currently two
+template <typename KernelName, typename KernelType>
+[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) { // expected-error 2 {{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
```

Comment:
Here. There are two errors triggered in this section- either the diagnostic is wrong, or the test cases which are expected to use the builtin correctly are somehow violating it.

Reply:
Can you please split the error-messages out like I'd mentioned with the VerifyDiagnosticConsumer's link (with the bookmarks) so that the errors are located in the location that instantiated them?

The code for this error is something I'm really quite confident in, but if you have an example that you think errors when it shouldn't, I can take a look.  If you run the testfile separately without -verify you can see where all the errors are with the associated notes.

Reply:
Yeah, I can't quite figure out which calls resolve to here to know what ones are triggering the unexpected errors. Putting the diagnostics markers next to the code that triggers the chain of diagnostics would really help track down what's going on.

Reply:
Ok. This is a struggle because every time I run llvm-lit, it results in different outputs showing that I missed adding diagnostic notes `// expected-note@+1{{in instantiation of function template specialization}}` However, I added all the diagnostic errors with number markers as suggested. Please take a look when you get a chance. Thank you!

Reply:
Maybe it's too early in the morning and I should be eating something with far too much sugar in it, but I have a few questions.

> This is a struggle because every time I run llvm-lit, it results in different outputs showing that I missed adding diagnostic notes

Are you saying that you're getting non-deterministic lit test failures with the code? Or do you mean that each time you ran the tests, you found another case where you missed converting an `expected-note@+N` to a bookmark?

> However, I added all the diagnostic errors with number markers as suggested.

So are the tests passing for you, or failing? If they're failing, can you provide the output (there's no CI being run for us to see how the tests are faring). I am guessing there are at least two cases failing unexpectedly, in `kernel13_14func()`.